### PR TITLE
feat: improve errors

### PR DESCRIPTION
**Pull request by fmvilas**
### Description
Improves the information provided by errors. It has several improvements:

1. Uses the Problem Details spec: https://tools.ietf.org/html/rfc7807. It makes it easier to differentiate error types.
1. Uses JSON Pointers to refer indicate where's the wrong field.
1. Adds the information about the line, the column, and the offset where the error happens. It's aware of the document format (YAML/JSON) and gives you this information correctly processed for the document format.
1. It makes it easier to differentiate between JSON and YAML documents.
1. Adds more tests.

##### Validation error

```js
{
  type: 'https://github.com/asyncapi/parser-js/validation-errors',
  title: 'There were errors validating the AsyncAPI document.',
  validationErrors: [{
    jsonPointer: '/components/schemas/testSchema/properties/test/$ref',
    startLine: 30,
    startColumn: 11,
    startOffset: 615,
    endLine: 30,
    endColumn: 34,
    endOffset: 638,
  }]
}
```

##### The document is not valid JSON nor YAML (YAML in this case)

```js
{
  type: 'https://github.com/asyncapi/parser-js/invalid-yaml,
  title: 'The provided YAML is not valid.',
  detail: 'duplicated mapping key at line 2, column -4:\n    bad:\n    ^',
  location: {
    jsonPointer: '/components/schemas/testSchema/properties/test/$ref',
    startLine: 2,
    startColumn: -4,
    startOffset: 5
  }
}
```

##### A $ref couldn't be resolved

```js
{
  type: 'https://github.com/asyncapi/parser-js/dereference-error',
  title: 'Error opening file "/Users/fmvilas/refs/refed.yaml" \nENOENT: no such file or directory, open '/Users/fmvilas/refs/refed.yaml'}',
  refs: [{
    jsonPointer: '/components/schemas/testSchema/properties/test/$ref',
    startLine: 30,
    startColumn: 11,
    startOffset: 615,
    endLine: 30,
    endColumn: 34,
    endOffset: 638,
  }]
}
```

### How to migrate

This PR introduces a breaking change. The error classes `PaserErrorNoJS` and `ParserErrorUnsupportedVersion` don't exist anymore. Instead, we only have a single `ParserError` class and you can guess the type of the error by looking at the `type` field of the it.

#### ParserErrorNoJS

**Before:**
```js
ParserErrorNoJS('Could not convert AsyncAPI to JSON.')
```

**Now:**
```js
ParserError({
  type: 'impossible-to-convert-to-json',
  title: 'Could not convert AsyncAPI to JSON.',
  detail: 'Most probably the AsyncAPI document contains invalid YAML or YAML features not supported in JSON.'
})
```

#### ParserErrorUnsupportedVersion

**Before:**
```js
ParserErrorUnsupportedVersion('AsyncAPI version is missing or unsupported: x.x.x.')
ParserErrorUnsupportedVersion.parsedJSON = {...}
```

**Now:**
```js
ParserError({
  type: 'missing-asyncapi-field',
  title: 'The `asyncapi` field is missing.',
  parsedJSON: {...},
}
```

and

```js
ParserError({
  type: 'unsupported-version',
  title: `Version x.x.x is not supported.`,
  detail: 'Please update to the latest version.',
  parsedJSON: {...},
  validationErrors: [{...}],
})
```

#### ParserError

**Before:**
```js
ParserError('many different error messages')
ParserError.parsedJSON = {...}
ParserError.errors = [{...}]
```

**Now:**

Different `ParserError` errors with different `type` fields to make it easier to identify the kind of message. For more information, see README.md.

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -1,27 +1,74 @@
...
+    throw new ParserError({
+      type: 'invalid-document-type',
+      title: 'The AsyncAPI document has to be either a string or a JS object.',
+    });
   }
 
-  try {
-    return JSON.parse(asyncapiYAMLorJSON);
-  } catch (e) {
+  if (asyncapiYAMLorJSON.trimLeft().startsWith('{')) {
```
Comment (fmvilas) [inline lib/utils.js:32]: This! 🤦 It is so easy to understand if a document is JSON or not. Don't know why I was trying a nested try/catch. Also, the nested try/catch had a problem: when it fails parsing json **and** yaml, you don't know if it's because it's invalid json or it's invalid yaml. It's impossible to guess the intent. The way it was, it was favoring YAML over JSON.

File: API.md [hunk 2]
```diff
@@ -1694,3 +1762,71 @@ Implements functions to deal with a Tag object.
...
+**Kind**: global function  
+
+| Param | Type | Description |
+| --- | --- | --- |
+| doc | [<code>AsyncAPIDocument</code>](#AsyncAPIDocument) |  |
+| callback(schema) | <code>function</code> | the function that is called foreach schema found.         schema {Schema}: the found schema. |
+
+<a name="assignIdToAnonymousSchemas"></a>
+
+## assignIdToAnonymousSchemas(doc)
```
Comment (derberg) [inline API.md:1824]: I see it is added here as it was not added in previous PRs. This will happen constantly.
What do you think if in a separate PR will add to release workflow a step with docs generation, so they get into npm and also get into the repository with package.json bump pr https://github.com/asyncapi/parser-js/blob/master/.github/workflows/release.yml#L41

Comment (fmvilas) [reply API.md:1824]: YES!

File: lib/parser.js [hunk 3]
```diff
@@ -32,50 +29,75 @@ module.exports = {
...
+        title: 'The `asyncapi` field is missing.',
+        parsedJSON,
+      });
+    }
+    
+    if (parsedJSON.asyncapi.startsWith('1.') || !asyncapi[parsedJSON.asyncapi]) {
+      throw new ParserError({
+        type: 'unsupported-version',
+        title: `Version ${parsedJSON.asyncapi} is not supported.`,
+        detail: 'Please upgrade to the latest version.',
```
[the remark this answers is not in the document]
Comment (fmvilas) [reply lib/parser.js:68]: 🤔 Maybe update? I don't want to explicitly mention the other tool here. The parser should live on its own with depending on the converters.

File: package.json [hunk 4]
```diff
@@ -43,10 +43,12 @@
     "semantic-release": "^17.0.4"
   },
   "dependencies": {
+    "@fmvilas/pseudo-yaml-ast": "^0.3.0",
```
[the remark this answers is not in the document]
Comment (fmvilas) [reply package.json:46]: Nope. It's here: https://github.com/fmvilas/pseudo-yaml-ast. It's not a fork by clicking on the fork button. I would like to keep evolving this tool separately.

File: lib/parser.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (derberg) [reply lib/parser.js]: sound fair, then maybe `Use latest version of the specification`

no upgrade, update or convert

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (derberg) [reply package.json:46]: oh, ok, I read it is fork in the readme and directly assumed it is physical fork

File: lib/parser.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (fmvilas) [reply lib/parser.js]: 👍

**Comment by fmvilas** [comment]
Thanks! 🚀
